Add tabs to organizational view

joeyyang opened the pull request:
Resolves #7526 

Now: 
![nov-01-2018 15-49-10](https://user-images.githubusercontent.com/2928395/47884374-b779ca80-dded-11e8-987b-3c13822bdad4.gif)

Before: 
![screenshot 2018-11-01 15 52 48](https://user-images.githubusercontent.com/2928395/47884505-3bcc4d80-ddee-11e8-8c44-f157fa81de16.png)

### Description
Adds tabs to the organizational view, separating cases into "Unassigned," "Assigned," and "Completed".

A few notes:
- Please see code comments in the "Files changed" view
- Please don't hold back if I've done anything that's not idiomatic in this codebase, bad style, etc. 
- Wasn't able to test this with more data besides the one case — is there a test case I can use that will have more data?
- This PR does not update the copy in the table, per https://github.com/department-of-veterans-affairs/caseflow/issues/7665. That will be addressed in a separate PR.

### Acceptance Criteria
- Not sure what goes here, can you please advise?

### Testing Plan
- Not sure what goes here, can you please advise?

joeyyang commented on client/app/queue/OrganizationQueue.jsx:
```diff
@@ -23,24 +29,30 @@ class OrganizationQueue extends React.PureComponent {
         <h1 {...fullWidth}>{sprintf(COPY.ORGANIZATION_QUEUE_TABLE_TITLE, this.props.organizationName)}</h1>
-        {content}
+        <TabWindow 
+          name="tasks-organization-queue"
```
Wasn't sure what to name this or if there was a convention — is this ok?

joeyyang commented on client/app/queue/OrganizationQueue.jsx:
```diff
@@ -63,3 +75,87 @@ const mapDispatchToProps = (dispatch) => ({
+    </React.Fragment>;
+  });
+
+const countCasesByType = (tasks) => {
```
Was wondering if there was a more elegant way to count up the # of cases in each category — any ideas for something better?

joeyyang commented on client/app/queue/OrganizationQueue.jsx:
```diff
@@ -5,42 +5,79 @@ import { bindActionCreators } from 'redux';
 import { fullWidth } from './constants';
 import COPY from '../../COPY.json';
 
+const countCasesByType = (tasks) => {
```
Was wondering if there was a more elegant way to count up the # of cases in each category — any ideas for something better?

lowellrex replied:
We can just get the length of the arrays returned by the selectors that you created in this PR (`getUnassignedOrganizationalTasks`, etc.).

For what it's worth, the mapping isn't 1:1 for tasks/cases, but right now we expect there to be only one task assigned to an organization for each case so this will work for now and be a very close approximation even when we start allowing the creation of multiple tasks for each case. Though we do need to keep this in mind as we move forward so we can create a more robust count when the time comes.

lowellrex commented on client/COPY.json:
```diff
@@ -88,6 +88,11 @@
+  "ORGANIZATIONAL_QUEUE_PAGE_UNASSIGNED_TAB_TITLE": "Unassigned (%d)",
+  "ORGANIZATIONAL_QUEUE_PAGE_ASSIGNED_TAB_TITLE": "Assigned (%d)",
+  "ORGANIZATIONAL_QUEUE_PAGE_COMPLETE_TAB_TITLE": "Completed",
+  "ORGANIZATIONAL_QUEUE_EMPTY_STATE_MESSAGE": "This queue is empty. You can ",
```
Can we add some copy here that explains what the three different tabs show? I think we can use the `COLOCATED_QUEUE_PAGE_COMPLETE_TASKS_DESCRIPTION` for the completed tab. For the "Unassigned" and "Assigned" tabs how does "Cases with tasks to be completed by the %s team that have not been assigned to any individual person" and "Cases assigned to a member of the %s team".

joeyyang replied:
Copy updated — lmk what you think!

lowellrex commented on client/app/queue/selectors.js:
```diff
@@ -99,6 +99,23 @@ export const getTasksForAppeal = createSelector(
 );
 
+export const getUnassignedOrganizationalTasks = createSelector(
+  [tasksWithAppealSelector],
```
`tasksWithAppealSelector()` returns all of the tasks in the global store regardless of who they are assigned to. Currently this only returns the tasks assigned to the organization by virtue of the fact that accessing the organizational queue table view forces a global state reload because we must directly navigate to their queue by typing in the organization's URL (or by doing some other browser-based navigation which will reload the webpage). However, should we ever change the way we navigate to the organizational queue we could have a global store with tasks for other cases not assigned to the organization.

This problem wasn't introduced in this PR, but I think we can handle this situation cleanly now by adding a selector that only picks out the organization's tasks here.

([Somewhat related ticket](https://github.com/department-of-veterans-affairs/caseflow/issues/7709) that aims to simplify how we manage task state between frontend and backend).

joeyyang replied:
created new selector in our pairing session — see updated code

lowellrex commented on client/app/queue/OrganizationQueue.jsx:
```diff
@@ -63,3 +100,75 @@ const mapDispatchToProps = (dispatch) => ({
+
+    return <React.Fragment>
+      {content}
+    </React.Fragment>;
```
Shouldn't be addressed in this PR, but I think the fact that we've got another place with task tables and tabs build the case for the [PR that consolidates that into a single component](https://github.com/department-of-veterans-affairs/caseflow/issues/7391).

lowellrex commented on client/app/queue/OrganizationQueue.jsx:
```diff
@@ -63,3 +100,75 @@ const mapDispatchToProps = (dispatch) => ({
+const CompletedTasksTab = connect(
+  (state: State) => ({ tasks: getCompletedOrganizationalTasks(state) }))(
+  (props: { tasks: Array<TaskWithAppeal> }) => {
+    const noTasks = !_.size(props.tasks);
```
I don't think we need to display text if there are not tasks for a certain tab. Eventually I think we'd like to show some text if there are no tasks at all for of the tabs, but until then (maybe #7391 makes this easy) I think it is fine to always show the table even if there are no rows in the table.

lowellrex commented on client/app/queue/selectors.js:
```diff
@@ -99,6 +99,23 @@ export const getTasksForAppeal = createSelector(
+  })
+);
+
+export const getAssignedOrganizationalTasks = createSelector(
```
Can we change the name of this function to something like `getOnHoldOrganizationalTasks`? Currently all tasks with the status of "on hold" should have a child task that is assigned to a member of that organization (or assigned to a different organization), but this might not always be the case.

lowellrex commented:
This looks really good! Left a few small comments that should be addressed but this is well on its way!

joeyyang commented:
This PR was changed to only get tasks matching the organization ID. Summary of changes (mostly just for my understanding of the codebase):
1. changed `tasks_controller.rb` to send back the organization ID from backend
2. change Redux stores to accept a notion of "active organization ID" to global state:
    1. change `uiReducer.js` to add `activeOrganizationId` field to global state
    2. add `SET_ACTIVE_ORGANIZATION_ID` to `uiConstants.js`
    3. add `setActiveOrganizationId` to `uiActions.js`
3. get that information in `OrganizationQueueLoadingScreen.jsx` — get it back from the response and then use the new `setActiveOrganizationId` action to set it as a prop in the global store
    1. also then bind that action creator to the global store
4. create a new selector which matches the current organization ID (passed in from above steps) to the global list of tasks, and display only those tasks

Updated screenshots — @lowellrex can you please review the copy?
<img width="1440" alt="screenshot 2018-11-05 12 35 31" src="https://user-images.githubusercontent.com/2928395/48025139-6a506e00-e0f7-11e8-9551-d4f4dba69b84.png">
<img width="1440" alt="screenshot 2018-11-05 12 35 40" src="https://user-images.githubusercontent.com/2928395/48025140-6a506e00-e0f7-11e8-9e53-59bec9735fa7.png">
<img width="1440" alt="screenshot 2018-11-05 12 35 44" src="https://user-images.githubusercontent.com/2928395/48025141-6a506e00-e0f7-11e8-9eaf-070d62190473.png">



New

lowellrex commented on client/COPY.json:
```diff
@@ -86,7 +86,14 @@
-  "COLOCATED_QUEUE_PAGE_COMPLETE_TASKS_DESCRIPTION": "Cases with tasks that were marked complete in the last two weeks will appeal here.",
+  "COLOCATED_QUEUE_PAGE_COMPLETE_TASKS_DESCRIPTION": "Cases with tasks that were marked complete in the last two weeks will appear here.",
+  "COLOCATED_QUEUE_PAGE_UNASSIGNED_TASKS_DESCRIPTION": "Cases owned by the %s team that are unassigned to a person.",
+  "COLOCATED_QUEUE_PAGE_ASSIGNED_TASKS_DESCRIPTION": "Cases assigned to a member of the %s team.",
```
I think it might make more sense to label these two pieces of copy with the prefix `ORGANIZATIONAL_QUEUE...` since they only appear on the org queue page for now.
